### PR TITLE
GH-2703: Added hideskipped option support for OpenCover

### PR DESCRIPTION
**Pull request by adison88**
- Added hideskipped support for OpenCover.
- Added unit tests.
- Added extension method to fetch all valid flag values.

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/cake-build/cake?pullRequest=2827) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/cake-build/cake?pullRequest=2827) before we can accept your contribution.<br/><br/>:x: adison88 [sign now](https://cla.dotnetfoundation.org/cake-build/cake?pullRequest=2827)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/cake-build/cake?pullRequest=2827) it.</sub>

File: src/Cake.Common/Tools/OpenCover/OpenCoverHideSkippedOption.cs [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    using System;
+
+    /// <summary>
+    /// Represents the hide skipped option for OpenCover.
+    /// </summary>
+    [Flags]
+    public enum OpenCoverHideSkippedOption
+    {
+        /// <summary>
+        /// Removes information that skipped due to switch <value>-excludebyfile </value> from output file.
```
Comment (Roemer) [inline src/Cake.Common/Tools/OpenCover/OpenCoverHideSkippedOption.cs:16]: Small typo in all of them, maybe rephrase to something like:
`Removes information from output files that relates to classes/modules that have been skipped (filtered) due to use of the <xxx>-switch.`
This would be in-line with the description in opencover.

File: src/Cake.Common/Tools/OpenCover/OpenCoverRunner.cs [hunk 2]
```diff
@@ -165,6 +165,13 @@ public sealed class OpenCoverRunner : Tool<OpenCoverSettings>
                 builder.AppendSwitch("-log", ":", settings.LogLevel.ToString());
             }
 
+            // HideSkipped Option
+            if (settings.HideSkippedOption != OpenCoverHideSkippedOption.All)
```
Comment (Roemer) [inline src/Cake.Common/Tools/OpenCover/OpenCoverRunner.cs:169]: I think the default is empty, so nothing gets skipped. You can then define the things that should be skipped and all seems to be just all of them. So we probably need a `HideSkippedOption.None` or `HideSkippedOption.Default` which you can use here to check.

File: src/Cake.Common/Tools/OpenCover/OpenCoverHideSkippedOption.cs [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+        /// <summary>
+        /// Removes missing Pdb information from output file.
+        /// </summary>
+        MissingPdb = 8,
+
+        /// <summary>
+        /// Removes all information from output file.
+        /// </summary>
+        All = 16
```
Comment (Roemer) [inline src/Cake.Common/Tools/OpenCover/OpenCoverHideSkippedOption.cs:38]: Shouldn't all be the combination of all of them? So like:
` All = File | Filter | Attribute | MissingPdb`

File: src/Cake.Common/Tools/OpenCover/OpenCoverSettings.cs [hunk 4]
```diff
@@ -115,6 +120,7 @@ public OpenCoverSettings()
             _excludedFileFilters = new HashSet<string>(StringComparer.Ordinal);
             Register = "user";
             LogLevel = OpenCoverLogLevel.Info;
+            HideSkippedOption = OpenCoverHideSkippedOption.All;
```
Comment (Roemer) [inline src/Cake.Common/Tools/OpenCover/OpenCoverSettings.cs:123]: Again, I think the default is empty/none?

File: src/Cake.Common/Tools/OpenCover/OpenCoverHideSkippedOption.cs (hunk 3 above)
Comment (adison88) [reply src/Cake.Common/Tools/OpenCover/OpenCoverHideSkippedOption.cs:38]: If I add like above then when it gets translated to string, it'll look like `-hideskipped:File;Filter;Attribute;MissingPdb` however it can take input as `-hideskipped:All`

File: src/Cake.Common/Tools/OpenCover/OpenCoverHideSkippedOption.cs (hunk 1 above)
Comment (adison88) [reply src/Cake.Common/Tools/OpenCover/OpenCoverHideSkippedOption.cs]: Updated.

File: src/Cake.Common/Tools/OpenCover/OpenCoverRunner.cs (hunk 2 above)
Comment (adison88) [reply src/Cake.Common/Tools/OpenCover/OpenCoverRunner.cs]: Added `None` option.

File: src/Cake.Common/Tools/OpenCover/OpenCoverSettings.cs (hunk 4 above)
Comment (adison88) [reply src/Cake.Common/Tools/OpenCover/OpenCoverSettings.cs]: Defaulted it to `None`

File: src/Cake.Common/Tools/OpenCover/OpenCoverHideSkippedOption.cs (hunk 3 above)
Comment (Roemer) [reply src/Cake.Common/Tools/OpenCover/OpenCoverHideSkippedOption.cs]: You could check in the runner if it is equal all and then replace by all. This would also replace it to All in case someone defines all individual flags.

**Comment by adison88** [comment]
@Roemer Updated as per the code review comments.

File: src/Cake.Common/Tools/OpenCover/OpenCoverHideSkippedOption.cs (hunk 3 above)
Comment (adison88) [reply src/Cake.Common/Tools/OpenCover/OpenCoverHideSkippedOption.cs]: I get it but I see that it also supports this `File;Filter;Attribute;MissingPdb;All`, do we need to append other options if All is appended already? If that's redundant then I can make the change.

Comment (adison88) [reply src/Cake.Common/Tools/OpenCover/OpenCoverHideSkippedOption.cs]: @Roemer I've updated the code as per the suggestion, Also added few more unit tests to cover `All` scenario, please check.

File: src/Cake.Common/Tools/OpenCover/OpenCoverHideSkippedOption.cs [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+    /// <summary>
+    /// Represents the hide skipped option for OpenCover.
+    /// </summary>
+    [Flags]
+    public enum OpenCoverHideSkippedOption
+    {
+        /// <summary>
+        /// Removes no information from output file.
+        /// </summary>
+        None = 1,
```
Comment (Roemer) [inline src/Cake.Common/Tools/OpenCover/OpenCoverHideSkippedOption.cs:18]: Just nitpicking ;) But I think in a flags scenario `None` has usually the value 0.

Comment (adison88) [reply src/Cake.Common/Tools/OpenCover/OpenCoverHideSkippedOption.cs]: Updated.

**Comment by Roemer** [comment]
lgtm 👍

**Comment by adison88** [comment]
@devlead Can you please review & merge?

**Comment by gep13** [comment]
@adison88 while we really appreciate you opening this PR, and doing the work to get this issue addressed, I want to make sure that we set realistic expectations of "when" this might be looked at...

Everyone on the Cake team works on Cake in their own spare time.  Almost everyone on the team was a family, and limited time that they can spend on open source.  This PR was only opened 2 days ago, which in the context of Open Source is around 2 minutes.  I would ask that you wait, patiently, for this to be reviewed, and hopefully, merged.

**Comment by adison88** [comment]
@gep13 Thank you for your comment. Being this my first contribution on Cake repository, I was not aware about the timeline for whole process to get PR in. No Worries and sorry for any inconvenience caused.

**Comment by gep13** [comment]
>@adison88 said...
I was not aware about the timeline for whole process to get PR in.

In my experience, there should be no expectations placed on _any_ Open Source project.

**Review by devlead — APPROVED** [review]
LGTM 👍 I'll merge if CI passes.

**Comment by devlead** [comment]
@adison88 your changes have been merged, thanks for your contribution 👍

**Comment by adison88** [comment]
Thanks @devlead for the merge.
